httpbakery/keyring.go

```diff
@@ -1,105 +1,113 @@
 	}
-	k, err := kr.cache.PublicKeyForLocation(loc)
+	info, err := kr.cache.ThirdPartyInfo(loc)
 	if err == nil {
-		return k, nil
+		return info, nil
 	}
-	k, err = PublicKeyForLocation(kr.client, loc)
+	info, err = ThirdPartyInfoForLocation(kr.client, loc)
 	if err != nil {
-		return nil, errgo.Mask(err)
-	}
-	if err := kr.cache.AddPublicKeyForLocation(loc, true, k); err != nil {
-		// Cannot happen in practice as it will only fail if
-		// loc is an invalid URL which we have already checked.
-		return nil, errgo.Notef(err, "cannot cache discharger URL %q", loc)
+		return bakery.ThirdPartyInfo{}, errgo.Mask(err)
 	}
-	return k, nil
+	kr.cache.AddInfo(loc, info)
+	return info, nil
 }
 
-// PublicKeyForLocation returns the public key from a macaroon
-// discharge server running at the given location URL.
-// Note that this is insecure if an http: URL scheme is used.
-// If client is nil, http.DefaultClient will be used.
-func PublicKeyForLocation(client *http.Client, url string) (*bakery.PublicKey, error) {
-	if client == nil {
-		client = http.DefaultClient
-	}
-	url = strings.TrimSuffix(url, "/") + "/publickey"
-	resp, err := client.Get(url)
+// ThirdPartyInfoForLocation returns information on the third party
+// discharge server running at the given location URL. Note that this is
+// insecure if an http: URL scheme is used. If client is nil,
+// http.DefaultClient will be used.
+func ThirdPartyInfoForLocation(client httprequest.Doer, url string) (bakery.ThirdPartyInfo, error) {
+	dclient := newDischargeClient(url, client)
+	var resp *http.Response
+	// We use Call directly instead of calling dclient.DischargeInfo because currently
+	// httprequest doesn't provide an easy way of finding out the HTTP response
+	// code from the error when the error response isn't marshaled as JSON.
+	// TODO(rog) change httprequest to make this straightforward.
+	err := dclient.Client.Call(&dischargeInfoRequest{}, &resp)
 	if err != nil {
-		return nil, errgo.Notef(err, "cannot get public key from %q", url)
-	}
-	if resp.StatusCode != http.StatusOK {
-		return nil, errgo.Newf("cannot get public key from %q: got status %s", url, resp.Status)
+		return bakery.ThirdPartyInfo{}, errgo.Mask(err)
 	}
 	defer resp.Body.Close()
-	data, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, errgo.Notef(err, "failed to read response body from %q", url)
-	}
-	var pubkey struct {
-		PublicKey *bakery.PublicKey
-	}
-	err = json.Unmarshal(data, &pubkey)
-	if err != nil {
-		return nil, errgo.Notef(err, "failed to decode response from %q", url)
+	var info bakery.ThirdPartyInfo
+	switch resp.StatusCode {
+	case http.StatusOK:
+		var r dischargeInfoResponse
+		if err := httprequest.UnmarshalJSONResponse(resp, &r); err != nil {
+			return bakery.ThirdPartyInfo{}, errgo.Mask(err)
+		}
+		info.PublicKey = *r.PublicKey
+		info.Version = r.Version
+	case http.StatusNotFound:
+		// TODO(rog) this fallback does not work because
```

Comment:
I trust you have relevant cards on your board to fix this and the above when the latest httprequest lands